### PR TITLE
[JENKINS-52785] Adds experimental multiarch Docker support

### PR DESCRIPTION
**Pull request by jmreicha**
This is a PoC I created to see if i could get multiarch Docker images building and working, it is still very rough and I think a lot of optimizations can be made.  The idea was to leave things currently working as is.

* Adds `publish-experimental.sh` script for building multiarch image without affecting current `publish.sh` script.
* `publish-experimental.sh` script is configured to push images to `jenkins/jenkins-experimental` Dockerhub registry.
* Creates new `multiarch` directory for storing multiarch Dockerfiles, leaving existing Dockerfiles alone.

[JENKINS-52785](https://issues.jenkins-ci.org/browse/JENKINS-52785)

### CODE REVIEW

**Comment by carlossg** [comment]
Would fix #686 #646

**Comment by jmreicha** [comment]
@carlossg @oleg-nenashev based on the conversation from the call earlier, it sounds like this maybe should get put on hold for now until the plan for creating different images under the `jenkins-experimental` org gets put into place?

**Comment by oleg-nenashev** [comment]
@jmreicha I propose to keep working on it. The org will be there soon, but we still need to have Dockerfiles stored somewhere. I do not mind if we keep them here even if the build runs elsewhere.

I rather have a concern about code duplication, which is an unfortunate issue with Docker since we cannot just parameterize image name in `FROM`

What if we create a Dockerfile generator in this repository and then use it to generate multi-arch Dockerfiles? It would be a bit tricky to get it integrated with DockerHub Autobuilds if we want, but for custom build/push flow it would save a lot of time.

The implementation could be just a simple sed/awk script.

**Comment by jmreicha** [comment]
Okay, good to know.  The Dockerfile duplication is one of the things I have been thinking how to improve so I will focus on that for now.

**Comment by jmreicha** [comment]
@imdurgadas As far as I understood from the call, there wasn't really much support for adding build agents for each architecture into the Jenkins infra?  If we can use native build agents, it would basically eliminate the need for qemu and all the extra Dockerfiles.

The Docker `manifest` command is still an experimental feature which is why I avoided using it.  There are lots of other projects using manifest-tool and it works well.

**Comment by jmreicha** [comment]
@imdurgadas PR has been updated to simplify the number of Dockerfiles.  You can use the image published on my dockerhub account if you want to test `jmreicha/jenkins:latest` or `jmreicha/jenkins:latest-s390x`.

There is a slight snafu with the s390x Alpine variant though, that I'm not sure how to handle.

```
Step 3/35 : RUN apk add --no-cache git openssh-client curl unzip bash ttf-dejavu coreutils tini
 ---> Running in 8ceacde0a70c
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/main/s390x/APKINDEX.tar.gz
WARNING: Ignoring http://dl-cdn.alpinelinux.org/alpine/v3.8/main/s390x/APKINDEX.tar.gz: IO ERROR
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/community/s390x/APKINDEX.tar.gz
WARNING: Ignoring http://dl-cdn.alpinelinux.org/alpine/v3.8/community/s390x/APKINDEX.tar.gz: IO ERROR
ERROR: unsatisfiable constraints:
  bash (missing):
    required by: world[bash]
  coreutils (missing):
    required by: world[coreutils]
  curl (missing):
    required by: world[curl]
  git (missing):
    required by: world[git]
  openssh-client (missing):
    required by: world[openssh-client]
  tini (missing):
    required by: world[tini]
  ttf-dejavu (missing):
    required by: world[ttf-dejavu]
  unzip (missing):
    required by: world[unzip]
  musl-1.1.19-r10:
    breaks:
            musl-utils-1.1.19-r10[musl?1.1.19-r10]
            musl-utils-1.1.19-r10[so:libc.musl-s390x.so.1]
            libbsd-0.8.6-r2[musl?1.1.16-r22]
            libbsd-0.8.6-r2[so:libc.musl-s390x.so.1]
            openjdk8-jre-base-8.171.11-r0[so:libc.musl-s390x.so.1]
            libxcb-1.13-r2[so:libc.musl-s390x.so.1]
            giflib-5.1.4-r2[so:libc.musl-s390x.so.1]
            busybox-1.28.4-r0[so:libc.musl-s390x.so.1]
            alsa-lib-1.1.6-r0[so:libc.musl-s390x.so.1]
            libxrender-0.9.10-r2[so:libc.musl-s390x.so.1]
            libtasn1-4.13-r0[so:libc.musl-s390x.so.1]
            ca-certificates-20171114-r3[so:libc.musl-s390x.so.1]
            lksctp-tools-1.0.17-r0[so:libc.musl-s390x.so.1]
            nss-3.36.1-r0[so:libc.musl-s390x.so.1]
            libxau-1.0.8-r2[so:libc.musl-s390x.so.1]
            libffi-3.2.1-r4[so:libc.musl-s390x.so.1]
            libxtst-1.2.3-r1[so:libc.musl-s390x.so.1]
            krb5-libs-1.15.2-r2[so:libc.musl-s390x.so.1]
            libressl2.7-libcrypto-2.7.4-r0[so:libc.musl-s390x.so.1]
            keyutils-libs-1.5.10-r0[so:libc.musl-s390x.so.1]
            lcms2-2.9-r0[so:libc.musl-s390x.so.1]
            p11-kit-0.23.10-r0[so:libc.musl-s390x.so.1]
            apk-tools-2.10.0-r0[so:libc.musl-s390x.so.1]
            libressl2.7-libssl-2.7.4-r0[so:libc.musl-s390x.so.1]
            scanelf-1.2.3-r0[so:libc.musl-s390x.so.1]
            zlib-1.2.11-r1[so:libc.musl-s390x.so.1]
            openjdk8-jre-8.171.11-r0[so:libc.musl-s390x.so.1]
            libpng-1.6.34-r1[so:libc.musl-s390x.so.1]
            nspr-4.19-r0[so:libc.musl-s390x.so.1]
            libcom_err-1.44.2-r0[so:libc.musl-s390x.so.1]
            libgcc-6.4.0-r8[so:libc.musl-s390x.so.1]
            libxcomposite-0.4.4-r1[so:libc.musl-s390x.so.1]
            libjpeg-turbo-1.5.3-r1[so:libc.musl-s390x.so.1]
            libx11-1.6.5-r1[so:libc.musl-s390x.so.1]
            alpine-baselayout-3.1.0-r0[so:libc.musl-s390x.so.1]
            p11-kit-trust-0.23.10-r0[so:libc.musl-s390x.so.1]
            pcsc-lite-libs-1.8.23-r1[so:libc.musl-s390x.so.1]
            libverto-0.3.0-r1[so:libc.musl-s390x.so.1]
            libxdmcp-1.1.2-r4[so:libc.musl-s390x.so.1]
            freetype-2.9.1-r1[so:libc.musl-s390x.so.1]
            openjdk8-8.171.11-r0[so:libc.musl-s390x.so.1]
            libxi-1.7.9-r1[so:libc.musl-s390x.so.1]
            libbz2-1.0.6-r6[so:libc.musl-s390x.so.1]
            sqlite-libs-3.24.0-r0[so:libc.musl-s390x.so.1]
            libstdc++-6.4.0-r8[so:libc.musl-s390x.so.1]
            libxext-1.3.3-r2[so:libc.musl-s390x.so.1]
  busybox-1.28.4-r0:
    breaks: world[busybox]
            ca-certificates-20171114-r3[/bin/sh]
            alpine-baselayout-3.1.0-r0[/bin/sh]
            java-common-0.1-r0[/bin/sh]
            java-cacerts-1.0-r0[/bin/sh]
  alpine-baselayout-3.1.0-r0:
    breaks: world[alpine-baselayout]
  alpine-keys-2.1-r1:
    breaks: world[alpine-keys]
  libressl2.7-libcrypto-2.7.4-r0:
    breaks:
            ca-certificates-20171114-r3[so:libcrypto.so.43]
            krb5-libs-1.15.2-r2[so:libcrypto.so.43]
            apk-tools-2.10.0-r0[so:libcrypto.so.43]
            libressl2.7-libssl-2.7.4-r0[so:libcrypto.so.43]
  libressl2.7-libssl-2.7.4-r0:
    breaks: krb5-libs-1.15.2-r2[so:libssl.so.45]
            apk-tools-2.10.0-r0[so:libssl.so.45]
  zlib-1.2.11-r1:
    breaks:
            openjdk8-jre-base-8.171.11-r0[so:libz.so.1]
            apk-tools-2.10.0-r0[so:libz.so.1]
            libpng-1.6.34-r1[so:libz.so.1]
            freetype-2.9.1-r1[so:libz.so.1]
  apk-tools-2.10.0-r0:
    breaks: world[apk-tools]
  scanelf-1.2.3-r0:
    breaks: musl-utils-1.1.19-r10[scanelf]
  musl-utils-1.1.19-r10:
    breaks: libc-utils-0.7.1-r0[musl-utils]
  libc-utils-0.7.1-r0:
    breaks: world[libc-utils]
  libffi-3.2.1-r4:
    breaks:
            openjdk8-jre-base-8.171.11-r0[so:libffi.so.6]
            p11-kit-0.23.10-r0[so:libffi.so.6]
  p11-kit-0.23.10-r0:
    breaks: java-cacerts-1.0-r0[p11-kit]
            p11-kit-trust-0.23.10-r0[so:libp11-kit.so.0]
  libtasn1-4.13-r0:
    breaks:
            p11-kit-trust-0.23.10-r0[so:libtasn1.so.6]
  p11-kit-trust-0.23.10-r0:
    breaks: java-cacerts-1.0-r0[p11-kit-trust]
  ca-certificates-20171114-r3:
    breaks: java-cacerts-1.0-r0[ca-certificates]
  java-cacerts-1.0-r0:
    breaks:
            openjdk8-jre-base-8.171.11-r0[java-cacerts]
            openjdk8-jre-8.171.11-r0[java-cacerts]
            openjdk8-8.171.11-r0[java-cacerts]
  libxau-1.0.8-r2:
    breaks: libxcb-1.13-r2[so:libXau.so.6]
  libbsd-0.8.6-r2:
    breaks: libxdmcp-1.1.2-r4[so:libbsd.so.0]
  libxdmcp-1.1.2-r4:
    breaks: libxcb-1.13-r2[so:libXdmcp.so.6]
  libxcb-1.13-r2:
    breaks: libx11-1.6.5-r1[so:libxcb.so.1]
  libx11-1.6.5-r1:
    breaks: libxrender-0.9.10-r2[so:libX11.so.6]
            libxtst-1.2.3-r1[so:libX11.so.6]
            openjdk8-jre-8.171.11-r0[so:libX11.so.6]
            libxcomposite-0.4.4-r1[so:libX11.so.6]
            libxi-1.7.9-r1[so:libX11.so.6]
            libxext-1.3.3-r2[so:libX11.so.6]
  libxcomposite-0.4.4-r1:
    breaks:
            openjdk8-jre-8.171.11-r0[so:libXcomposite.so.1]
  libxext-1.3.3-r2:
    breaks: libxtst-1.2.3-r1[so:libXext.so.6]
            openjdk8-jre-8.171.11-r0[so:libXext.so.6]
            libxi-1.7.9-r1[so:libXext.so.6]
  libxi-1.7.9-r1:
    breaks:
            openjdk8-jre-8.171.11-r0[so:libXi.so.6]
  libxrender-0.9.10-r2:
    breaks:
            openjdk8-jre-8.171.11-r0[so:libXrender.so.1]
  libxtst-1.2.3-r1:
    breaks:
            openjdk8-jre-8.171.11-r0[so:libXtst.so.6]
  alsa-lib-1.1.6-r0:
    breaks:
            openjdk8-jre-8.171.11-r0[so:libasound.so.2]
  libbz2-1.0.6-r6:
    breaks: freetype-2.9.1-r1[so:libbz2.so.1]
  libpng-1.6.34-r1:
    breaks:
            openjdk8-jre-8.171.11-r0[so:libpng16.so.16]
            freetype-2.9.1-r1[so:libpng16.so.16]
  freetype-2.9.1-r1:
    breaks:
            openjdk8-jre-8.171.11-r0[so:libfreetype.so.6]
  libgcc-6.4.0-r8:
    breaks: nss-3.36.1-r0[so:libgcc_s.so.1]
            openjdk8-jre-8.171.11-r0[so:libgcc_s.so.1]
            libstdc++-6.4.0-r8[so:libgcc_s.so.1]
  giflib-5.1.4-r2:
    breaks:
            openjdk8-jre-8.171.11-r0[so:libgif.so.7]
  libjpeg-turbo-1.5.3-r1:
    breaks:
            openjdk8-jre-base-8.171.11-r0[so:libjpeg.so.8]
            openjdk8-jre-8.171.11-r0[so:libjpeg.so.8]
  libstdc++-6.4.0-r8:
    breaks:
            openjdk8-jre-base-8.171.11-r0[so:libstdc++.so.6]
            nss-3.36.1-r0[so:libstdc++.so.6]
            openjdk8-jre-8.171.11-r0[so:libstdc++.so.6]
  openjdk8-jre-lib-8.171.11-r0:
    breaks:
            openjdk8-jre-base-8.171.11-r0[openjdk8-jre-lib]
  java-common-0.1-r0:
    breaks:
            openjdk8-jre-base-8.171.11-r0[java-common]
  krb5-conf-1.0-r1:
    breaks: krb5-libs-1.15.2-r2[krb5-conf]
  libcom_err-1.44.2-r0:
    breaks:
            krb5-libs-1.15.2-r2[so:libcom_err.so.2]
  keyutils-libs-1.5.10-r0:
    breaks:
            krb5-libs-1.15.2-r2[so:libkeyutils.so.1]
  libverto-0.3.0-r1:
    breaks: krb5-libs-1.15.2-r2[so:libverto.so.1]
  krb5-libs-1.15.2-r2:
    breaks:
            openjdk8-jre-base-8.171.11-r0[so:libkrb5.so.3]
  lcms2-2.9-r0:
    breaks:
            openjdk8-jre-base-8.171.11-r0[so:liblcms2.so.2]
  nspr-4.19-r0:
    breaks:
            openjdk8-jre-base-8.171.11-r0[so:libnspr4.so]
            nss-3.36.1-r0[so:libnspr4.so]
            nss-3.36.1-r0[so:libplc4.so]
            nss-3.36.1-r0[so:libplds4.so]
  sqlite-libs-3.24.0-r0:
    breaks: nss-3.36.1-r0[so:libsqlite3.so.0]
  nss-3.36.1-r0:
    breaks:
            openjdk8-jre-base-8.171.11-r0[so:libnss3.so]
            openjdk8-jre-base-8.171.11-r0[so:libnssutil3.so]
  pcsc-lite-libs-1.8.23-r1:
    breaks:
            openjdk8-jre-base-8.171.11-r0[so:libpcsclite.so.1]
  lksctp-tools-1.0.17-r0:
    breaks:
            openjdk8-jre-base-8.171.11-r0[so:libsctp.so.1]
  openjdk8-jre-base-8.171.11-r0:
    breaks:
            openjdk8-jre-8.171.11-r0[so:openjdk8:libawt.so]
            openjdk8-jre-8.171.11-r0[so:openjdk8:libjava.so]
            openjdk8-jre-8.171.11-r0[so:openjdk8:libjli.so]
            openjdk8-jre-8.171.11-r0[so:openjdk8:libjvm.so]
            openjdk8-8.171.11-r0[so:openjdk8:libjli.so]
  openjdk8-jre-8.171.11-r0:
    breaks:
            openjdk8-jre-8.171.11-r0[openjdk8-jre]
            openjdk8-8.171.11-r0[openjdk8-jre]
            openjdk8-8.171.11-r0[so:openjdk8:libawt_xawt.so]
  openjdk8-8.171.11-r0:
    breaks: world[openjdk8?8.171.11-r0]
The command '/bin/sh -c apk add --no-cache git openssh-client curl unzip bash ttf-dejavu coreutils tini' returned a non-zero code: 59
```

**Comment by imdurgadas** [comment]
@jmreicha , Pulled the image on s390x successfully !

```
root@z-ub-001:~# uname -a
Linux z-ub-001 4.4.0-134-generic #160-Ubuntu SMP Wed Aug 15 14:55:40 UTC 2018 s390x s390x s390x GNU/Linux

root@z-ub-001:~# docker run --name jenkins-s390x -P jmreicha/jenkins:latest
Running from: /usr/share/jenkins/jenkins.war
webroot: EnvVars.masterEnvVars.get("JENKINS_HOME")
Sep 03, 2018 4:32:51 AM org.eclipse.jetty.util.log.Log initialized
INFO: Logging initialized @11448ms to org.eclipse.jetty.util.log.JavaUtilLog
....

```

**Comment by jmreicha** [comment]
@imdurgadas Would you mind trying to build the Alpine variant on your s390x system and see if you get similar errors to mine?

**Comment by imdurgadas** [comment]
Sure @jmreicha . I built alpine on s390x successfully.

```
root@z-ub-001:~/durgadas/docker# docker build -t jenkins-alphine -f Dockerfile-alpine .
Sending build context to Docker daemon  669.2kB
Step 1/34 : FROM openjdk:8-jdk-alpine
8-jdk-alpine: Pulling from library/openjdk
cdf21ace9418: Pull complete
ea178433f2f0: Pull complete
b948ae8f2138: Pull complete
9de7d6fd0a5e: Pull complete
Digest: sha256:1fd5a77d82536c88486e526da26ae79b6cd8a14006eb3da3a25eb8d2d682ccd6
Status: Downloaded newer image for openjdk:8-jdk-alpine
 ---> c9bbe1a3aa09
Step 2/34 : RUN apk add --no-cache git openssh-client curl unzip bash ttf-dejavu coreutils tini
 ---> Running in c583e43c932c
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/main/s390x/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/community/s390x/APKINDEX.tar.gz
(1/25) Installing ncurses-terminfo-base (6.1_p20180818-r0)
(2/25) Installing ncurses-terminfo (6.1_p20180818-r0)
(3/25) Installing ncurses-libs (6.1_p20180818-r0)
(4/25) Installing readline (7.0.003-r0)
(5/25) Installing bash (4.4.19-r1)
Executing bash-4.4.19-r1.post-install
(6/25) Installing libattr (2.4.47-r7)
(7/25) Installing libacl (2.2.52-r5)
(8/25) Installing coreutils (8.29-r2)
(9/25) Installing nghttp2-libs (1.32.0-r0)
(10/25) Installing libssh2 (1.8.0-r3)
(11/25) Installing libcurl (7.61.0-r0)
(12/25) Installing curl (7.61.0-r0)
(13/25) Installing expat (2.2.5-r0)
(14/25) Installing pcre2 (10.31-r0)
(15/25) Installing git (2.18.0-r0)
(16/25) Installing openssh-keygen (7.7_p1-r3)
(17/25) Installing openssh-client (7.7_p1-r3)
(18/25) Installing tini (0.18.0-r0)
(19/25) Installing fontconfig (2.12.6-r1)
(20/25) Installing encodings (1.0.4-r1)
(21/25) Installing libfontenc (1.1.3-r2)
(22/25) Installing mkfontscale (1.1.3-r0)
(23/25) Installing mkfontdir (1.0.7-r1)
(24/25) Installing ttf-dejavu (2.37-r0)
(25/25) Installing unzip (6.0-r4)
Executing busybox-1.28.4-r0.trigger
Executing fontconfig-2.12.6-r1.trigger
Executing mkfontscale-1.1.3-r0.trigger
Executing mkfontdir-1.0.7-r1.trigger
OK: 149 MiB in 78 packages
Removing intermediate container c583e43c932c
 ---> ed514af8bf7e
Step 3/34 : ARG user=jenkins
 ---> Running in 7be87e08cdd4
Removing intermediate container 7be87e08cdd4
 ---> 81d781ed2ee1
Step 4/34 : ARG group=jenkins
 ---> Running in 47b06e9059cc
Removing intermediate container 47b06e9059cc
 ---> a48d61ef67d0
Step 5/34 : ARG uid=1000
 ---> Running in 8892fcf11c0a
Removing intermediate container 8892fcf11c0a
 ---> c244de6cefad
Step 6/34 : ARG gid=1000
 ---> Running in 13689359567e
Removing intermediate container 13689359567e
 ---> 1db207a9e266
Step 7/34 : ARG http_port=8080
 ---> Running in b0bdb411cc7e
Removing intermediate container b0bdb411cc7e
 ---> c647aabc546d
Step 8/34 : ARG agent_port=50000
 ---> Running in 7645f72005be
Removing intermediate container 7645f72005be
 ---> 891b1db4e528
Step 9/34 : ARG JENKINS_HOME=/var/jenkins_home
 ---> Running in dab37f3ddfc3
Removing intermediate container dab37f3ddfc3
 ---> 5a2b84846f44
Step 10/34 : ENV JENKINS_HOME $JENKINS_HOME
 ---> Running in ed4df902021f
Removing intermediate container ed4df902021f
 ---> d859e6f399da
Step 11/34 : ENV JENKINS_SLAVE_AGENT_PORT ${agent_port}
 ---> Running in f13f7f57718b
Removing intermediate container f13f7f57718b
 ---> 44e1bcae586e
Step 12/34 : RUN mkdir -p $JENKINS_HOME   && chown ${uid}:${gid} $JENKINS_HOME   && addgroup -g ${gid} ${group}   && adduser -h "$JENKINS_HOME" -u ${uid} -G ${group} -s /bin/bash -D ${user}
 ---> Running in 6da2c15579e0
Removing intermediate container 6da2c15579e0
 ---> 7039cf148b88
Step 13/34 : VOLUME $JENKINS_HOME
 ---> Running in e55a731b6560
Removing intermediate container e55a731b6560
 ---> 3e3d268cf229
Step 14/34 : RUN mkdir -p /usr/share/jenkins/ref/init.groovy.d
 ---> Running in 4f8b099798b1
Removing intermediate container 4f8b099798b1
 ---> a64065b097d1
Step 15/34 : COPY init.groovy /usr/share/jenkins/ref/init.groovy.d/tcp-slave-agent-port.groovy
 ---> b3f0c7e3b219
Step 16/34 : ARG JENKINS_VERSION
 ---> Running in bd4f396ac937
Removing intermediate container bd4f396ac937
 ---> 98250f6895b7
Step 17/34 : ENV JENKINS_VERSION ${JENKINS_VERSION:-2.60.3}
 ---> Running in 8178f6195d8a
Removing intermediate container 8178f6195d8a
 ---> 45e5540899c1
Step 18/34 : ARG JENKINS_SHA=2d71b8f87c8417f9303a73d52901a59678ee6c0eefcf7325efed6035ff39372a
 ---> Running in e870dabd815a
Removing intermediate container e870dabd815a
 ---> 651dd623404a
Step 19/34 : ARG JENKINS_URL=https://repo.jenkins-ci.org/public/org/jenkins-ci/main/jenkins-war/${JENKINS_VERSION}/jenkins-war-${JENKINS_VERSION}.war
 ---> Running in 2d37faef8105
Removing intermediate container 2d37faef8105
 ---> 56f0b9991b5b
Step 20/34 : RUN curl -fsSL ${JENKINS_URL} -o /usr/share/jenkins/jenkins.war   && echo "${JENKINS_SHA}  /usr/share/jenkins/jenkins.war" | sha256sum -c -
 ---> Running in 0a56a0331f9b
/usr/share/jenkins/jenkins.war: OK
Removing intermediate container 0a56a0331f9b
 ---> d7599f3cb58e
Step 21/34 : ENV JENKINS_UC https://updates.jenkins.io
 ---> Running in 85d23e238d8b
Removing intermediate container 85d23e238d8b
 ---> 711987644980
Step 22/34 : ENV JENKINS_UC_EXPERIMENTAL=https://updates.jenkins.io/experimental
 ---> Running in 860b151f6924
Removing intermediate container 860b151f6924
 ---> b7e14813b64b
Step 23/34 : ENV JENKINS_INCREMENTALS_REPO_MIRROR=https://repo.jenkins-ci.org/incrementals
 ---> Running in a1664cc0f850
Removing intermediate container a1664cc0f850
 ---> 1dfcd819b931
Step 24/34 : RUN chown -R ${user} "$JENKINS_HOME" /usr/share/jenkins/ref
 ---> Running in 857d0522643b
Removing intermediate container 857d0522643b
 ---> 49c7a3eeeb6a
Step 25/34 : EXPOSE ${http_port}
 ---> Running in bb581775f62b
Removing intermediate container bb581775f62b
 ---> 71c3a421ee7c
Step 26/34 : EXPOSE ${agent_port}
 ---> Running in 79628db712f3
Removing intermediate container 79628db712f3
 ---> 5bc783e292c4
Step 27/34 : ENV COPY_REFERENCE_FILE_LOG $JENKINS_HOME/copy_reference_file.log
 ---> Running in 31a2340b5289
Removing intermediate container 31a2340b5289
 ---> c7b851dfa00d
Step 28/34 : USER ${user}
 ---> Running in 6d60b166fa10
Removing intermediate container 6d60b166fa10
 ---> 2a2ab944699f
Step 29/34 : COPY jenkins-support /usr/local/bin/jenkins-support
 ---> 5acfae76e83b
Step 30/34 : COPY jenkins.sh /usr/local/bin/jenkins.sh
 ---> 61caa29e96d7
Step 31/34 : COPY tini-shim.sh /bin/tini
 ---> 7a23a62321a3
Step 32/34 : ENTRYPOINT ["/sbin/tini", "--", "/usr/local/bin/jenkins.sh"]
 ---> Running in 1edcb4721556
Removing intermediate container 1edcb4721556
 ---> 381bd287fd01
Step 33/34 : COPY plugins.sh /usr/local/bin/plugins.sh
 ---> 7388cced80b2
Step 34/34 : COPY install-plugins.sh /usr/local/bin/install-plugins.sh
 ---> 0e201c51285e
Successfully built 0e201c51285e
Successfully tagged jenkins-alphine:latest

```

**Comment by jmreicha** [comment]
@imdurgadas Interesting.  The s390x Alpine variant seems to be the only build causing me problems, and I'm not really sure what the problem is at this point.

**Comment by jmreicha** [comment]
@imdurgadas Good idea, just pushed.

**Comment by imdurgadas** [comment]
@jmreicha on s390x , the cross compile script worked well for alpine. However on amd64, the script failed for s390x/alpine with same error as you pasted above (https://github.com/jenkinsci/docker/pull/719#issuecomment-417958985)

```
Step 1/34 : FROM openjdk:8-jdk-alpine
8-jdk-alpine: Pulling from library/openjdk
Digest: sha256:1fd5a77d82536c88486e526da26ae79b6cd8a14006eb3da3a25eb8d2d682ccd6
Status: Image is up to date for openjdk:8-jdk-alpine
 ---> c9bbe1a3aa09
Step 2/34 : RUN apk add --no-cache git openssh-client curl unzip bash ttf-dejavu coreutils tini
 ---> Running in 80d4516bbd5e
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/main/s390x/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/community/s390x/APKINDEX.tar.gz
(1/25) Installing ncurses-terminfo-base (6.1_p20180818-r0)
(2/25) Installing ncurses-terminfo (6.1_p20180818-r0)
(3/25) Installing ncurses-libs (6.1_p20180818-r0)
(4/25) Installing readline (7.0.003-r0)
(5/25) Installing bash (4.4.19-r1)

```

**Comment by imdurgadas** [comment]
@jmreicha it's working fine now with the QEMU version bump.

**Comment by oleg-nenashev** [comment]
@jmreicha As discussed after the Platform SIG meeting, I have created https://hub.docker.com/r/jenkins4eval/jenkins-multiarch-quemu and added you as a contributor. After going through the code, this approach seems to be almost ready to go.

**Comment by imdurgadas** [comment]
@jmreicha Can you also add `ppc64le` in QEMUARCHS and below condition in `set-base-image` function
```
elif [[ $arch == ppc64le ]]; then
        BASEIMAGE="ppc64le/openjdk:8-jdk"
fi
```

@oleg-nenashev What's your thought on adding `Jenkinsfile-experimental` file so that we can configure it to push it to jenkins4eval repo ?

**Comment by oleg-nenashev** [comment]
> What's your thought on adding Jenkinsfile-experimental file so that we can configure it to push it to jenkins4eval repo ?

Fine with me (note that such approach may cause difficulties for Multi-branch builds). In such case I would expect a section in README describing the experimental support.

**Comment by oleg-nenashev** [comment]
🤦‍♂️ , fixed: https://hub.docker.com/r/jenkins4eval/jenkins-multiarch-qemu

**Comment by jmreicha** [comment]
@oldelvet PR has been updated.  Let me know if there's anything to do.

**Comment by oleg-nenashev** [comment]
Just heads-up, I will be adding images for OpenJDK 11 soon. Would be nice to get this PR merged first so that we prevent collisions. I would also like to move the BASEIMAGE approach to the upstream so we can avoid massive duplication of Dockerfile contents.

P.S: I wonder why Docker still does not offer a built-in templating engine for builds

**Comment by jmreicha** [comment]
Would be happy to do another PR for adding the base image method to
upstream after this one is merged.

**Comment by oleg-nenashev** [comment]
@ndeloof any objections?

File: publish-experimental.sh [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+    if [[ $variant =~ alpine && $arch == arm ]]; then
+        BASEIMAGE="arm32v6/openjdk:8-jdk-alpine"
+    elif [[ $variant =~ alpine ]]; then
+        BASEIMAGE="$BASEIMAGE-alpine"
+    elif [[ $variant =~ slim ]]; then
+        BASEIMAGE="$BASEIMAGE-slim"
+    fi
+
+    # Make the Dockerfile after we set the base image
+    sed -i "s|BASEIMAGE|${BASEIMAGE}|g" "$dockerfile"
```
Comment (oleg-nenashev) [inline publish-experimental.sh]: It won't work well on Darwin where there is no `-i` flag.
I have a for of this script with a fix, can submit later

**Comment by oleg-nenashev** [comment]
@carlossg @ndeloof Would you be fine if I go forward with that?

**Comment by oleg-nenashev** [comment]
Merged. @jmreicha would you be also able to create some documentation in README or in other file about this flow?

**Comment by imdurgadas** [comment]
@oleg-nenashev can we also add additional line in `Jenkinsfile` to trigger publish-experimental.sh ?

**Comment by oleg-nenashev** [comment]
@imdurgadas yes, if it is generic enough. Do you want to run it on another instance?

**Comment by jmreicha** [comment]
@oleg-nenashev I just opened https://github.com/jenkinsci/docker/pull/750 to add docs

**Pull request merged**